kv/dist_sender.go

```diff
@@ -480,6 +469,132 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
+	// back to us within a reasonable amount of time).
+	// We may simply not be trying to talk to the up-to-date replicas, so
+	// clearing the descriptor here should be a good idea.
+	// TODO(tschottdorf): If a replica group goes dead, this will cause clients
+	// to put high read pressure on the first range, so there should be some
+	// rate limiting here.
+	if rpcErr != nil {
+		if log.V(1) {
+			log.Warningf("RPC error; expiring cached descriptor")
+		}
+		ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+	}
+
+	if err := util.FirstError(rpcErr, reply.Header().GoError()); err != nil {
+		if log.V(1) {
+			log.Warningf("failed to invoke %s: %s", call.Method(), err)
+		}
+
+		// If retryable, allow retry. For range not found or range
+		// key mismatch errors, we don't backoff on the retry,
+		// but reset the backoff loop so we can retry immediately.
+		switch tErr := err.(type) {
+		case *proto.RangeNotFoundError, *proto.RangeKeyMismatchError:
+			// Range descriptor might be out of date - evict it.
+			ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+			// On addressing errors, don't backoff; retry immediately.
+			return retry.Reset, nil
+		case *proto.NotLeaderError:
+			newLeader := tErr.GetLeader()
+			// Verify that leader is a known replica; if not,
+			// we've got a stale replica; evict cache.
+			// Next, cache the new leader.
+			if newLeader != nil {
+				if i := replicas.FindReplica(newLeader.StoreID); i == -1 {
+					ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+				}
+			} else {
+				newLeader = &proto.Replica{}
+			}
+			ds.updateLeaderCache(proto.RaftID(desc.RaftID), *newLeader)
+			return retry.Reset, nil
+		default:
+			if retryErr, ok := err.(util.Retryable); ok && retryErr.CanRetry() {
+				return retry.Continue, nil
+			}
+		}
+		return retry.Break, err
+	}
+	return retry.Break, nil
+}
+
+// getDescriptors takes a call and looks up the corresponding range descriptors
+// associated to it. First, the range descriptor for the corresponding StartKey
+// from call.Args is returned; second, if the EndKey exceeds that of the that
+// descriptor, its descriptor is looked up as well.
+func (ds *DistSender) getDescriptors(call client.Call) (*proto.RangeDescriptor, *proto.RangeDescriptor, error) {
+	// If this is an InternalPushTxn, set ignoreIntents option as
+	// necessary. This prevents a potential infinite loop; see the
+	// comments in proto.InternalRangeLookupRequest.
+	options := lookupOptions{}
+	if pushArgs, ok := call.Args.(*proto.InternalPushTxnRequest); ok {
+		options.ignoreIntents = pushArgs.RangeLookup
+	}
+
+	desc, err := ds.rangeCache.LookupRangeDescriptor(call.Args.Header().Key, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var descNext *proto.RangeDescriptor
+	// If the request accesses keys beyond the end of this range,
+	// get the descriptor of the adjacent range to address next.
+	if desc.EndKey.Less(call.Args.Header().EndKey) {
```

Comment:
I think we need the adjacent range if these are equal, too, since `desc.EndKey` is exclusive